### PR TITLE
Update Debug

### LINKED ISSUE #1118: Debug doesn't work when you first do upgrade
When debugging the first time you upgrade a container is created and doesn't pause on breakpoints

### PR DESCRIPTION
**Pull request by cazfletch**
kill the chaincode container if running
move instantiate and upgrade to when play is pressed
update readme

closes #1118

Signed-off-by: Caroline Fletcher <caroline.fletcher@uk.ibm.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,10 +83,16 @@ To debug your smart contract follow these steps:
...
-6. Add breakpoints to the smart contract by clicking on the relevant line numbers in your smart contract files.
-7. To submit or evaluate a transaction, click the blockchain icon on the debug toolbar. Alternatively, in the `Fabric Gateways` panel, you can right click on transactions to submit or evaluate them. Execution will be paused on any breakpoints you've defined.
+4. Select the `Debug Smart Contract` configuration by using the dropdown in the upper-left and click the **play** button on the debug toolbar. The extension will automatically instantiate or upgrade the smart contract as appropriate. If you want to test out a function that is called on upgrade add the following to the launch configuration Where name is the name of the smart contract and version is different to the previous version used. Alternatively if you are using Node then you can update the version in the package.json file.
```
Comment (Jakeeyturner) [inline README.md]: "If you want to test out a function that is called on **instantiate or upgrade** add the following to the launch configuration, **where 'name'** is the name of the smart contract and 'version' is different to the previous version used"

File: README.md [hunk 2]
```diff
@@ -83,12 +83,18 @@ To debug your smart contract follow these steps:
...
-6. Add breakpoints to the smart contract by clicking on the relevant line numbers in your smart contract files.
-7. To submit or evaluate a transaction, click the blockchain icon on the debug toolbar. Alternatively, in the `Fabric Gateways` panel, you can right click on transactions to submit or evaluate them. Execution will be paused on any breakpoints you've defined.
+4. Select the `Debug Smart Contract` configuration by using the dropdown in the upper-left and click the **play** button on the debug toolbar. The extension will automatically instantiate or upgrade the smart contract as appropriate. If you want to test out a function that is called on instantiate or upgrade add the following to the launch configuration, where `name` is the name of the smart contract and `version` is different to the previous version used. Alternatively if you are using Node then you can update the `version` in the package.json file.
```
Comment (Jakeeyturner) [inline README.md]: Just noticed `Alternatively if you are using Node` - should we change this to be clearer `Alternatively, if you are debugging a JavaScript or TypeScript smart contract...`

File: README.md [hunk 3]
```diff
@@ -83,12 +83,18 @@ To debug your smart contract follow these steps:
...
+```
+    "env": {
+        "CORE_CHAINCODE_ID_NAME": <name>:<version>
+    }    
+```
+
+5. Add breakpoints to the smart contract by clicking on the relevant line numbers in your smart contract files.
+6. To submit or evaluate a transaction, click the blockchain icon on the debug toolbar. Alternatively, in the `Fabric Gateways` panel, you can right click on transactions to submit or evaluate them. Execution will be paused on any breakpoints you've defined.
+
+To make iterative changes to your smart contract while debugging, after making your changes click the **restart** button. You can also stop the debugging session, make further changes and start debugging again, without needing to upgrade your smart contract.
```
Comment (Jakeeyturner) [inline README.md]: Should we add a subheading like 'Hot Reloading' so people are more likely to know that this is a feature?

File: client/src/debug/FabricDebugConfigurationProvider.ts [hunk 4]
```diff
@@ -36,7 +36,7 @@ export abstract class FabricDebugConfigurationProvider implements vscode.DebugCo
             const extensionData: ExtensionData = context.globalState.get<ExtensionData>(EXTENSION_DATA_KEY);
 
             // Stop debug if not got late enough version
-            if (!extensionData.generatorVersion || semver.lt(extensionData.generatorVersion, '0.0.33')) {
+            if (!extensionData.generatorVersion || semver.lt(extensionData.generatorVersion, '0.0.36')) {
                 outputAdapter.log(LogType.ERROR, 'To debug a smart contract, you must update the local Fabric runtime. Teardown and start the local Fabric runtime, and try again.', 'To debug a smart contract, you must update the local Fabric runtime. Teardown and start the local Fabric runtime, and try again.');
```
Comment (Jakeeyturner) [inline client/src/debug/FabricDebugConfigurationProvider.ts]: Pretty sure I wrote this line 👀, but can we make it:
```
outputAdapter.log(LogType.ERROR, 'To debug a smart contract, you must update the local Fabric runtime. Teardown and start the local Fabric runtime, and try again.');
```

File: client/src/debug/FabricGoDebugConfigurationProvider.ts [hunk 5]
```diff
@@ -29,9 +29,10 @@ export class FabricGoDebugConfigurationProvider extends FabricDebugConfiguration
         ];
     }
 
-    protected async getChaincodeName(): Promise<string> {
+    protected async getChaincodeNameAndVersion(): Promise<{name: string, version: string}> {
         const name: string = await UserInputUtil.showInputBox('Enter a name for your Go package'); // Getting the specified name and package from the user
-        return name;
+        const version: string = await UserInputUtil.showInputBox('Enter a version for your Go package');
```
Comment (Jakeeyturner) [inline client/src/debug/FabricGoDebugConfigurationProvider.ts]: Should we not be checking if the user has cancelled giving us a name or version?

Comment (cazfletch) [reply client/src/debug/FabricGoDebugConfigurationProvider.ts]: this is checked in the FabricDebugConfigurationProvider

**Pull request merged**